CONTRIBUTING.md

```diff
@@ -1,8 +1,90 @@
-<!-- TODO: Section on nix CI + devShell -->
+# Contributing guide
 
-## Running tests locally
+Contributions are more than welcome!
 
-For reproducibility, we only run tests that can be sandboxed in CI.
+## Commit messages / PR title
+
+Please ensure your pull request title conforms to [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
+
+## CI
+
+Our CI checks are run using [`nix`](https://nixos.org/download.html#download-nix).
+
+## Development
+
+To be able to reproduce CI, we recommend using `nix`.
+
+### Dev environment
+
+We use the following tools:
+
+#### Formatting
+
+- [`rustfmt`](https://github.com/rust-lang/rustfmt) [Rust]
+- [`alejandra`](https://github.com/kamadorueda/alejandra) [Nix]
+
+#### Linting
+
+- [`cargo check`](https://doc.rust-lang.org/cargo/commands/cargo-check.html)
+- [`clippy`](https://doc.rust-lang.org/clippy/)
+
+### Nix devShell
+
+- Requires [flakes](https://nixos.wiki/wiki/Flakes) to be enabled.
+
+We provide a `flake.nix` that can bootstrap all of the above development tools.
+
+To enter a development shell:
+
+```console
+nix develop
+```
+
+To apply formatting, while in a devShell, run
+
+```console
+pre-commit run --all
+```
+
+If you use [`direnv`](https://direnv.net/),
+just run `direnv allow` and you will be dropped in this devShell.
+
+### Running tests
+
+The easiest way to run tests is with Nix (see below).
+
+If you do not use Nix, you can also run the test suite locally.
+
+### Running tests and checks with Nix
+
+If you just want to run all checks that are available, run:
+
+```console
+nix flake check -Lv --accept-flake-config
+```
+
+To run individual checks, using Nix:
+
+```console
+nix build .#checks.<your-system>.<check> -Lv --accept-flake-config
```

Comment:
the `--accept-flake-config` makes the command harder to use and I dont think it's an issue for a contributor to answer "Y" the first time.